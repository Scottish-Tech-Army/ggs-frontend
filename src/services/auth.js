const BASE_URL = process.env.REACT_APP_AWS_CLIENT_API_ENDPOINT;

export function login(email) {
  return fetch(BASE_URL + "unit/login", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ email }),
  }).then(response => {
    if (response.ok) {
      return response.json();
    }
    throw response;
  })
}

export function register(email, name) {
  return fetch(BASE_URL + "unit/register", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ email, name }),
  }).then(response => {
    if (response.ok) {
      return response.json();
    }
    throw response;
  })
}
