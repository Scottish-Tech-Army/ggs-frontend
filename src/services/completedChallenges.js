const BASE_URL = process.env.REACT_APP_AWS_CLIENT_API_ENDPOINT;


// NOTE: in the original code the following function 
// was called getLeaderboard, hence the use of 
// "unit/leaderboard" in the first arg of fetch()
// below.
export function getCompletedChallenges(email) {
  return fetch(BASE_URL + "unit/leaderboard", {
    method: "GET",
    headers: {
      GGSUnit: email,
    },
  })
    .then((data) => data.json()) // the .json() returns the actual data
    .catch((err) => {
      console.error(err);
    });
}
