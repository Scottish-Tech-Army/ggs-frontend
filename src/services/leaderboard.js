const BASE_URL = process.env.REACT_APP_AWS_CLIENT_API_ENDPOINT;

export function getLeaderboard(email) {
  return fetch(BASE_URL + "unit/leaderboard", {
    method: "GET",
    headers: {
      GGSUnit: email,
    },
  })
    .then((data) => data.json())
    .catch((err) => {
      console.error(err);
    });
}
