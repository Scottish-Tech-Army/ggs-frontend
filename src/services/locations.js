const BASE_URL = process.env.REACT_APP_AWS_CLIENT_API_ENDPOINT;

export function getLocations(email) {
  return fetch(BASE_URL + "locations", {
    method: "GET",
    headers: {
      GGSUnit: email,
    },
  })
    .then((data) => data.json())
    .catch((err) => {
      console.error(err);
    });
}

export function collectLocation(email, id) {
  return fetch(BASE_URL + "unit/collect", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      GGSUnit: email,
    },
    body: JSON.stringify({ id })
  }).catch((err) => {
    console.error(err);
  });
}
