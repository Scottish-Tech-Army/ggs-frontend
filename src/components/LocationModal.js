import React, { useContext, useEffect, useState } from "react";
import Modal from "react-bootstrap/Modal";
import Button from "react-bootstrap/Button";
import Image from "react-bootstrap/Image";
import { collectLocation } from "../services/locations";
import { authContext } from "../contexts/AuthContext";
import dividerLine from "./divider-line.svg";
import xPrimary from "./x-primary.svg";
import placeholderPhoto from "./image-coming-soon.png";

// Change this as needed for coordinate distance from landmark.
const LOCATION_TOLERANCE_LATITUDE = 0.0002; // Note 0.0002 is approx equal to 22 metres.
const LOCATION_TOLERANCE_LONGITUDE = 0.0004; // Note 0.0004 is approx equal to 24 metres.

const isLocationInRange = (location, userLatLong) => {
  return (
    Math.abs(location.latitude - userLatLong.latitude) <= LOCATION_TOLERANCE_LATITUDE &&
    Math.abs(location.longitude - userLatLong.longitude) <= LOCATION_TOLERANCE_LONGITUDE
  );
};

const LocationModal = ({
  selectedLocation,
  handleCloseLocation,
  setLocations,
  userLatLong,
}) => {
  const { unit } = useContext(authContext);

  const [isOutOfRange, setIsOutOfRange] = useState(true);
  const [deviceErrMsg, setDeviceErrMsg] = useState();

  const handleCollectLocation = (event) => {
    collectLocation(unit.email, selectedLocation.locationId)
      .then((response) => {
        if (response.ok) {
          // Update the frontend locations with the collected marker to match the backend status
          setLocations((locations) => {
            const index = locations.findIndex(
              (i) => i.locationId === selectedLocation.locationId
            );
            const locationList = [...locations];
            locationList[index].collected = true;
            return locationList;
          });
        }
      })
      .catch((error) => {
        console.error(error);
      });
    event.preventDefault();
  };

  // Mukund: Set a variable to the object that is in array photos 
  // of object selectedLocation (location). array photos looks like this:
  // [{url:  "https://dz1ex1yb0vxyz.cloudfront.net/dunbartonshire-bearsden-thegruffalotrail.jpg"}]
    const photo = selectedLocation.photos[0];

  useEffect(() => {
    if (!userLatLong) {
      setDeviceErrMsg("Please turn on location tracking");
      setIsOutOfRange(true);
    } else {
      setIsOutOfRange(!isLocationInRange(selectedLocation, userLatLong));
    }
  }, [selectedLocation, userLatLong]);

  // Adapt the collecting button depending on user's proximity to the landmark location
  // and the collection status of the landmark location
  const collectButtonDisabled = selectedLocation.collected || isOutOfRange;
  const collectButtonText = isOutOfRange
    ? "Please come closer to this location"
    : "Tap to reveal challenge";

  const areaName =
    selectedLocation.city && selectedLocation.city !== selectedLocation.county
      ? `${selectedLocation.city}, ${selectedLocation.county}`
      : selectedLocation.county;

  function getPhoto(photo) {
    if (!photo) {
      return (
        <>
          <Image
            className="img-location"
            src={placeholderPhoto}
            alt="no image available"
            rounded
          />
          <div className="challenge">
            <h2>Bonus Challenge</h2>
            <div className="content">
              Will you be the first to take a picture of this location? Take a
              photo and send it to us at{" "}
              <a href="mailto:web@girlguiding-scot.org.uk">
                web@girlguiding-scot.org.uk
              </a>
            </div>
          </div>
        </>
      );
    }

    let creditLine = null;
    if (photo.attribution || photo.copyright) {
      let attributionElement = photo.originalUrl ? (
        <a href={photo.originalUrl} target="_blank" rel="noreferrer">
          {photo.attribution} {photo.copyright}
        </a>
      ) : (
        <>
          {photo.attribution} {photo.copyright}
        </>
      );
      creditLine = (
        <div className="img-location-credit">Credit: {attributionElement}</div>
      );
    }

    return (
      <>
        <Image
          className="img-location"
          src={photo.url}
          alt={selectedLocation.name}
          rounded
        />
        {creditLine}
      </>
    );
  }

  return (
    <Modal
      show={true}
      onHide={handleCloseLocation}
      key={selectedLocation.locationId}
      className="custom-modal location-modal"
    >
      <Modal.Header className="border-0 mb-n4">
        <Button
          variant="outline-primary"
          onClick={handleCloseLocation}
          className="closer-position"
          aria-label="Close"
        >
          <img
            src={xPrimary}
            style={{ width: "200%", height: "200%" }}
            alt=""
          />
        </Button>
      </Modal.Header>
      <Modal.Body className="mt-n3">
        <div className="place-name">{selectedLocation.name}</div>
        <div className="city-name">{areaName}</div>
        <div className="scroll-container">
          <div className="scroll">
            {getPhoto(photo)}
            <div className="description">{selectedLocation.description}</div>
            {selectedLocation.collected && (
              <div className="challenge">
                <h2>Challenge</h2>
                <div className="content">{selectedLocation.challenge}</div>
              </div>
            )}
          </div>
        </div>
      </Modal.Body>
      {!!deviceErrMsg && (
        <div className="container">
          <img src={dividerLine} style={{ width: "100%" }} alt="" />
          <p className="feedback-branding">{deviceErrMsg}</p>
        </div>
      )}
      {!selectedLocation.collected && !deviceErrMsg && (
        <Button
          bsPrefix="btn-branding"
          onClick={handleCollectLocation}
          disabled={collectButtonDisabled}
          className={
            collectButtonDisabled
              ? "btn-branding-disabled mx-2 mb-2"
              : "btn-branding-enabled mx-2 mb-2"
          }
        >
          {collectButtonText}
        </Button>
      )}
    </Modal>
  );
};

export default LocationModal;
