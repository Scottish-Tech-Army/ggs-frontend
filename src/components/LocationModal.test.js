/* eslint-disable testing-library/no-node-access */
import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import LocationModal from "./LocationModal";
import { collectLocation } from "../services/locations";
import AuthProvider from "../contexts/AuthContext";

const handleCloseLocation = jest.fn();
const setLocations = jest.fn();

jest.mock("../services/locations");
const localStorageGetItem = jest.spyOn(
  Object.getPrototypeOf(window.localStorage),
  "getItem"
);
const localStorageSetItem = jest.spyOn(
  Object.getPrototypeOf(window.localStorage),
  "setItem"
);

describe("component LocationModal", () => {
  const TEST_LOCATION = {
    city: "Edinburgh",
    photos: [{ url: "https://localhost/photo.jpg" }],
    challenge: "This is the challenge text",
    county: "Edinburgh",
    longitude: -3.21329494646239,
    locationId: "edinburgh-edinburgh-testlocation",
    region: "Edinburgh",
    description: "This is the location description.",
    latitude: 55.94846683803023,
    name: "Test location name",
  };

  const OTHER_LOCATION = {
    city: "Edinburgh2",
    photos: [{ url: "https://localhost/photo.jpg" }],
    challenge: "This is the challenge text2",
    county: "Edinburgh2",
    longitude: -3.21329494646239,
    locationId: "edinburgh-edinburgh-testlocation2",
    region: "Edinburgh2",
    description: "This is the location description2.",
    latitude: 55.94846683803023,
    name: "Test location name2",
  };

  const LOCATION_WITHOUT_PHOTO = {
    city: "Edinburgh",
    photos: [],
    challenge: "This is the challenge text",
    county: "Edinburgh",
    longitude: -3.21329494646239,
    locationId: "edinburgh-edinburgh-testlocation",
    region: "Edinburgh",
    description: "This is the location description.",
    latitude: 55.94846683803023,
    name: "Test location name",
  };

  const IN_RANGE_LOCATION = { longitude: -3.213294, latitude: 55.948466 };
  const OUT_OF_RANGE_LOCATION = { longitude: -3.213294, latitude: 55.949 };

  const TEST_EMAIL = "test@example.com";

  beforeEach(() => {
    localStorageGetItem.mockReturnValue(
      JSON.stringify({
        email: TEST_EMAIL,
        name: "Test team",
      })
    );
  });

  afterAll(() => {
    localStorageGetItem.mockRestore();
    localStorageSetItem.mockRestore();
  });

  function renderWithUser(component) {
    return {
      user: userEvent.setup(),
      ...render(<AuthProvider>{component}</AuthProvider>),
    };
  }

  it("show location information", () => {
    renderWithUser(
      <LocationModal
        selectedLocation={TEST_LOCATION}
        handleCloseLocation={handleCloseLocation}
        setLocations={setLocations}
        userLatLong={IN_RANGE_LOCATION}
      />
    );

    const modal = document.querySelector(".modal-dialog");
    expect(modal).toMatchSnapshot();
    expect(modal).toHaveTextContent("Test location name");
    expect(modal).toHaveTextContent("Edinburgh");
    expect(modal).toHaveTextContent("This is the location description.");
  });

  it("show out of range location", () => {
    renderWithUser(
      <LocationModal
        selectedLocation={TEST_LOCATION}
        handleCloseLocation={handleCloseLocation}
        setLocations={setLocations}
        userLatLong={OUT_OF_RANGE_LOCATION}
      />
    );

    const modal = document.querySelector(".modal-dialog");

    const button = screen.getByRole("button", {
      name: "Please come closer to this location",
    });
    expect(button).toBeDisabled();
    expect(modal).not.toHaveTextContent("This is the challenge text");
  });

  it("show in range uncollected location", () => {
    renderWithUser(
      <LocationModal
        selectedLocation={TEST_LOCATION}
        handleCloseLocation={handleCloseLocation}
        setLocations={setLocations}
        userLatLong={IN_RANGE_LOCATION}
      />
    );

    const modal = document.querySelector(".modal-dialog");

    const button = screen.getByRole("button", { name: "Tap to reveal challenge" });
    expect(button).not.toBeDisabled();
    expect(modal).not.toHaveTextContent("This is the challenge text");
  });

  it("show in range collected location", () => {
    renderWithUser(
      <LocationModal
        selectedLocation={{ ...TEST_LOCATION, collected: true }}
        handleCloseLocation={handleCloseLocation}
        setLocations={setLocations}
        userLatLong={IN_RANGE_LOCATION}
      />
    );

    const modal = document.querySelector(".modal-dialog");
    expect(modal).toHaveTextContent("This is the challenge text");

    expect(
      screen.queryByRole("button", { name: "Tap to reveal challenge" })
    ).toBeNull();
  });

  it("collect location", async () => {
    collectLocation.mockResolvedValue({ ok: true });

    const { user } = renderWithUser(
      <LocationModal
        selectedLocation={TEST_LOCATION}
        handleCloseLocation={handleCloseLocation}
        setLocations={setLocations}
        userLatLong={IN_RANGE_LOCATION}
      />
    );

    await user.click(screen.getByRole("button", { name: "Tap to reveal challenge" }));

    expect(collectLocation).toHaveBeenCalledTimes(1);
    expect(collectLocation).toHaveBeenCalledWith(
      TEST_EMAIL,
      "edinburgh-edinburgh-testlocation"
    );

    expect(setLocations).toHaveBeenCalledTimes(1);
    const setLocationsFilter = setLocations.mock.calls[0][0];
    expect(
      setLocationsFilter([OTHER_LOCATION, TEST_LOCATION, OTHER_LOCATION])
    ).toEqual([
      OTHER_LOCATION,
      { ...TEST_LOCATION, collected: true },
      OTHER_LOCATION,
    ]);
  });

  it("placeholder image", async () => {
    collectLocation.mockResolvedValue({ ok: true });

    renderWithUser(
      <LocationModal
        selectedLocation={LOCATION_WITHOUT_PHOTO}
        handleCloseLocation={handleCloseLocation}
        setLocations={setLocations}
        userLatLong={IN_RANGE_LOCATION}
      />
    );

    const modal = document.querySelector(".modal-dialog");

    expect(screen.getByAltText("no image available")).toHaveAttribute(
      "src",
      "image-coming-soon.png"
    );
    expect(modal).toHaveTextContent("Take a photo");
    expect(screen.getByRole("link")).toHaveAttribute(
      "href",
      "mailto:web@girlguiding-scot.org.uk"
    );
  });

  it("fire close on clicking close button", async () => {
    const { user } = renderWithUser(
      <LocationModal
        selectedLocation={TEST_LOCATION}
        handleCloseLocation={handleCloseLocation}
        setLocations={setLocations}
        userLatLong={IN_RANGE_LOCATION}
      />
    );

    await user.click(screen.getByLabelText("Close"));

    expect(handleCloseLocation).toHaveBeenCalledTimes(1);
  });
});
