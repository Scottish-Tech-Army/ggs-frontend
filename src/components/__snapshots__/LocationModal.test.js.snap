// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component LocationModal show location information 1`] = `
<div
  class="modal-dialog"
>
  <div
    class="modal-content"
  >
    <div
      class="border-0 mb-n4 modal-header"
    >
      <button
        aria-label="Close"
        class="closer-position btn btn-outline-primary"
        type="button"
      >
        <img
          alt=""
          src="x-primary.svg"
          style="width: 200%; height: 200%;"
        />
      </button>
    </div>
    <div
      class="mt-n3 modal-body"
    >
      <div
        class="place-name"
      >
        Test location name
      </div>
      <div
        class="city-name"
      >
        Edinburgh
      </div>
      <div
        class="scroll-container"
      >
        <div
          class="scroll"
        >
          <img
            alt="Test location name"
            class="img-location rounded"
            src="https://localhost/photo.jpg"
          />
          <div
            class="description"
          >
            This is the location description.
          </div>
        </div>
      </div>
    </div>
    <button
      class="btn-branding-enabled mx-2 mb-2 btn-branding btn-branding-primary"
      type="button"
    >
      Tap to reveal challenge
    </button>
  </div>
</div>
`;
